freqtrade/exchange/exchange.py

```diff
@@ -1704,6 +1746,164 @@ def set_margin_mode(self, pair: str, collateral: Collateral, params: dict = {}):
+
+        fees: float = 0
+        if not close_date:
+            close_date = datetime.now(timezone.utc)
+        open_timestamp = int(open_date.timestamp()) * 1000
+        # close_timestamp = int(close_date.timestamp()) * 1000
+        funding_rate_history = self.get_funding_rate_history(
+            pair,
+            open_timestamp
+        )
+        mark_price_history = self._get_mark_price_history(
+            pair,
+            open_timestamp
+        )
+        funding_fee_dates = self._get_funding_fee_dates(open_date, close_date)
+        for date in funding_fee_dates:
+            timestamp = int(date.timestamp()) * 1000
+            if timestamp in funding_rate_history:
+                funding_rate = funding_rate_history[timestamp]
+            else:
+                logger.warning(
+                    f"Funding rate for {pair} at {date} not found in funding_rate_history"
+                    f"Funding fee calculation may be incorrect"
+                )
+            if timestamp in mark_price_history:
+                mark_price = mark_price_history[timestamp]
+            else:
+                logger.warning(
+                    f"Mark price for {pair} at {date} not found in funding_rate_history"
+                    f"Funding fee calculation may be incorrect"
+                )
+            if funding_rate and mark_price:
+                fees += self._get_funding_fee(
+                    size=amount,
+                    mark_price=mark_price,
+                    funding_rate=funding_rate
+                )
+
+        return fees
+
+    def get_funding_fees(self, pair: str, amount: float, open_date: datetime):
+        """
+        Fetch funding fees, either from the exchange (live) or calculates them
+        based on funding rate/mark price history
+        :param pair: The quote/base pair of the trade
+        :param amount: Trade amount
+        :param open_date: Open date of the trade
+        """
+        if self._config['dry_run']:
+            funding_fees = self._calculate_funding_fees(pair, amount, open_date)
+        else:
+            funding_fees = self._get_funding_fees_from_exchange(pair, open_date)
+        return funding_fees
+
+    @retrier
+    def get_funding_rate_history(self, pair: str, since: int) -> Dict:
+        """
+        :param pair: quote/base currency pair
+        :param since: timestamp in ms of the beginning time
+        :param end: timestamp in ms of the end time
+        """
+        if not self.exchange_has("fetchFundingRateHistory"):
+            raise ExchangeError(
+                f"CCXT has not implemented fetchFundingRateHistory for {self.name}; "
+                f"therefore, dry-run/backtesting for {self.name} is currently unavailable"
+            )
+
+        # TODO-lev: Gateio has a max limit into the past of 333 days, okex has a limit of 3 months
+        try:
+            funding_history: Dict = {}
+            response = self._api.fetch_funding_rate_history(
```

Comment:
I wanted to write a "live" test for this endpoint ... but it doesn't seem to be working (ccxt==1.60.92, installed 20 minutes ago).

```python
import ccxt
ccxt.binanceusdm()
exchange.fetch_funding_rate_history('BTC/USDT', '1h', limit=100)
```

fails with 
```
ExchangeError: binanceusdm {"status":"ERROR","type":"VALID","code":"100001003","errorData":{"startTime":"Failed to convert property value of type 'java.lang.String' to required type 'java.lang.Long' for property 'startTime'; nested exception is java.lang.NumberFormatException: For input string: \"1h\""},"data":null,"subData":null,"params":null}
```

Reply:
> I wanted to write a "live" test for this endpoint ... but it doesn't seem to be working (ccxt==1.60.92, installed 20 minutes ago).
> 
> ```python
> import ccxt
> ccxt.binanceusdm()
> exchange.fetch_funding_rate_history('BTC/USDT', '1h', limit=100)
> ```
> 
> fails with
> 
> ```
> ExchangeError: binanceusdm {"status":"ERROR","type":"VALID","code":"100001003","errorData":{"startTime":"Failed to convert property value of type 'java.lang.String' to required type 'java.lang.Long' for property 'startTime'; nested exception is java.lang.NumberFormatException: For input string: \"1h\""},"data":null,"subData":null,"params":null}
> ```

This is the method header for `fetch_funding_rate_history`, you're passing `1h` into the argument `since`

```
def fetch_funding_rate_history(self, symbol=None, since=None, limit=None, params={}):
```

Reply:
mmmmh ... good point :laughing: